File: sdk/java/src/main/java/network/ferrum/sdk/FerrumClient.java
package network.ferrum.sdk;

import java.math.BigInteger;

import org.bouncycastle.crypto.digests.Blake2bDigest;

import io.emeraldpay.polkaj.api.PolkadotApi;
import io.emeraldpay.polkaj.api.RpcCall;
import io.emeraldpay.polkaj.api.StandardCommands;
import io.emeraldpay.polkaj.apiws.JavaHttpSubscriptionAdapter;
import io.emeraldpay.polkaj.schnorrkel.Schnorrkel;
import io.emeraldpay.polkaj.types.Address;
import io.emeraldpay.polkaj.types.Hash256;
import io.emeraldpay.polkaj.ss58.SS58Type;

import network.ferrum.sdk.Calls.FerrumCall;

/**
 * Typed thin wrapper over polkaj for the Ferrum sovereign blockchain.
 *
 * polkaj provides the WebSocket transport, sr25519 (schnorrkel) signing and SS58
 * codec; this client SCALE-encodes the call params ({@link Calls}) and assembles a
 * v4 signed extrinsic with the runtime's 8-field SignedExtra (immortal era, zero
 * tip), then submits it via {@code author_submitExtrinsic}.
 */
public final class FerrumClient implements AutoCloseable {
    public static final String DEFAULT_ENDPOINT = "ws://127.0.0.1:9944";

    public final Calls.Identity identity = new Calls.Identity();
    public final Calls.Credential credential = new Calls.Credential();
    public final Calls.Tax tax = new Calls.Tax();
    public final Calls.Treasury treasury = new Calls.Treasury();
    public final Calls.Federation federation = new Calls.Federation();
    public final Calls.Interop interop = new Calls.Interop();

    private final PolkadotApi api;
    private final JavaHttpSubscriptionAdapter adapter;

    private FerrumClient(PolkadotApi api, JavaHttpSubscriptionAdapter adapter) {
        this.api = api;
        this.adapter = adapter;
    }

    public PolkadotApi raw() { return api; }

    public static FerrumClient connect(String endpoint) throws Exception {
        var adapter = JavaHttpSubscriptionAdapter.newBuilder().connectTo(endpoint).build();
        var api = PolkadotApi.newBuilder().subscriptionAdapter(adapter).build();
        adapter.connect().get();
        return new FerrumClient(api, adapter);
    }

    /** Build a sr25519 keypair from a 32-byte seed (e.g. a wallet seed). */
    public static Schnorrkel.KeyPair keypairFromSeed(byte[] seed32) throws Exception {
        return Schnorrkel.getInstance().generateKeyPairFromSeed(seed32);
    }

    /** Sign and submit a Ferrum call; returns the extrinsic hash. */
    public Hash256 signAndSend(FerrumCall call, Schnorrkel.KeyPair signer) throws Exception {
        byte[] pub = signer.getPublicKey();
        byte[] callBytes = call.encode();

        var rv = api.execute(StandardCommands.getInstance().getRuntimeVersion()).get();
        Hash256 genesis = api.execute(StandardCommands.getInstance().getBlockHash(0)).get();
        String ss58 = new Address(SS58Type.Network.SUBSTRATE, pub).toString();
        Integer nonce = api.execute(RpcCall.create(Integer.class, "system_accountNextIndex", ss58)).get();

        byte[] extra = new ScaleWriter()
                .u8(0)                                   // CheckEra: immortal
                .compact(BigInteger.valueOf(nonce))      // CheckNonce
                .compact(BigInteger.ZERO)                // ChargeTransactionPayment tip
                .toArray();
        byte[] additional = new ScaleWriter()
                .u32(rv.getSpecVersion())                // CheckSpecVersion
                .u32(rv.getTransactionVersion())         // CheckTxVersion
                .fixed(genesis.getBytes())               // CheckGenesis
                .fixed(genesis.getBytes())               // CheckEra (immortal -> genesis)
                .toArray();

        byte[] payload = concat(callBytes, extra, additional);
        if (payload.length > 256) payload = blake2b256(payload);
        byte[] sig = Schnorrkel.getInstance().sign(payload, signer);

        byte[] body = new ScaleWriter()
                .u8(0x84)            // v4, signed
                .u8(0x00).fixed(pub) // MultiAddress::Id(account)
                .u8(0x01).fixed(sig) // MultiSignature::Sr25519
                .fixed(extra)
                .fixed(callBytes)
                .toArray();

        byte[] framed = new ScaleWriter().bytes(body).toArray(); // compact-length prefix
        String hex = "0x" + toHex(framed);
        return api.execute(RpcCall.create(Hash256.class, "author_submitExtrinsic", hex)).get();
    }

    @Override
    public void close() {
        api.close();
    }

    // ---- helpers ----

    private static byte[] concat(byte[]... parts) {
        int n = 0;
        for (byte[] p : parts) n += p.length;
        byte[] out = new byte[n];
        int o = 0;
        for (byte[] p : parts) { System.arraycopy(p, 0, out, o, p.length); o += p.length; }
        return out;
    }

    private static byte[] blake2b256(byte[] data) {
        Blake2bDigest d = new Blake2bDigest(256);
        d.update(data, 0, data.length);
        byte[] out = new byte[32];
        d.doFinal(out, 0);
        return out;
    }

    private static String toHex(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) sb.append(String.format("%02x", x));
        return sb.toString();
    }
}

File: sdk/csharp/Ferrum.Sdk/FerrumClient.cs
using System.Numerics;
using Substrate.NetApi;
using Substrate.NetApi.Model.Extrinsics;
using Substrate.NetApi.Model.Rpc;
using Substrate.NetApi.Model.Types;

namespace Ferrum.Sdk;

/// <summary>
/// Typed thin wrapper over Substrate.NetApi's <see cref="SubstrateClient"/>.
///
/// The SDK encodes Ferrum call params itself (see <see cref="Calls"/>) and assembles
/// the signed extrinsic by hand to match the runtime's 8-field SignedExtra. We do NOT
/// delegate to Substrate.NetApi's extrinsic builder: that version appends a
/// CheckMetadataHash extension byte the Ferrum runtime does not carry, which shifts the
/// call and traps validate_transaction ("Bad input data").
/// </summary>
public sealed class FerrumClient : IAsyncDisposable
{
    public const string DefaultEndpoint = "ws://127.0.0.1:9944";

    public SubstrateClient Api { get; }
    public IdentityCalls Identity { get; } = new();
    public CredentialCalls Credential { get; } = new();
    public TaxCalls Tax { get; } = new();
    public TreasuryCalls Treasury { get; } = new();
    public FederationCalls Federation { get; } = new();
    public InteropCalls Interop { get; } = new();

    private FerrumClient(SubstrateClient api) => Api = api;

    public static async Task<FerrumClient> ConnectAsync(string endpoint = DefaultEndpoint, CancellationToken token = default)
    {
        var client = new SubstrateClient(new Uri(endpoint), ChargeTransactionPayment.Default());
        // useMetaData + standardSubstrate so RuntimeVersion and GenesisHash are populated.
        await client.ConnectAsync(true, true, token);
        return new FerrumClient(client);
    }

    /// <summary>Build a sr25519 account from a 64-byte expanded secret + 32-byte public key.</summary>
    public static Account Account(byte[] secretKey, byte[] publicKey) =>
        Substrate.NetApi.Model.Types.Account.Build(KeyType.Sr25519, secretKey, publicKey);

    /// <summary>
    /// Build a sr25519 account from a 32-byte secret seed (the "Secret seed" printed by
    /// <c>ferrum-node key generate</c>). The seed is expanded to the 64-byte key Schnorrkel
    /// signs with and the public key is derived — passing the raw seed to <see cref="Account"/>
    /// throws SignatureError::ByteLengthError.
    /// </summary>
    public static Account AccountFromSeed(byte[] seed) =>
        Substrate.NetApi.Model.Types.Account.FromSeed(KeyType.Sr25519, seed);

    /// <summary>Sign and submit a Ferrum call; returns the extrinsic hash.</summary>
    /// <remarks>
    /// Assembles a v4 signed extrinsic with the runtime's 8-field SignedExtra
    /// (immortal era, zero tip). <paramref name="lifeTime"/> is accepted for API
    /// compatibility but unused — the era is immortal, so the era checkpoint hash is
    /// the genesis hash.
    /// </remarks>
    public async Task<string> SignAndSendAsync(FerrumCall call, Account account, uint lifeTime = 64, CancellationToken token = default)
    {
        _ = lifeTime;
        var callBytes = call.Encode();

        var rv = Api.RuntimeVersion;
        var genesis = Api.GenesisHash.Bytes;
        var nonce = Convert.ToUInt32(await Api.System.AccountNextIndexAsync(account.Value, token));

        // extra (the signed part): immortal era ++ compact(nonce) ++ compact(tip = 0)
        var extra = new ScaleWriter().U8(0).Compact(nonce).Compact(BigInteger.Zero).ToArray();
        // additional signed: specVersion ++ txVersion ++ genesis ++ era checkpoint (== genesis)
        var additional = new ScaleWriter()
            .U32(rv.SpecVersion).U32(rv.TransactionVersion).Fixed(genesis).Fixed(genesis).ToArray();

        var payload = new ScaleWriter().Fixed(callBytes).Fixed(extra).Fixed(additional).ToArray();
        if (payload.Length > 256) payload = HashExtension.Blake2(payload, 256, null); // sign the hash for long payloads
        var signature = await account.SignAsync(payload);

        var body = new ScaleWriter()
            .U8(0x84).U8(0x00).Fixed(account.Bytes) // v4 signed, MultiAddress::Id(public key)
            .U8(0x01).Fixed(signature)              // MultiSignature::Sr25519
            .Fixed(extra).Fixed(callBytes)
            .ToArray();

        var framed = new ScaleWriter().Bytes(body).ToArray(); // compact length prefix ++ body
        var hash = await Api.Author.SubmitExtrinsicAsync(Utils.Bytes2HexString(framed), token);
        return Utils.Bytes2HexString(hash.Bytes);
    }

    /// <summary>Subscribe to all chain events via the System.Events storage key.</summary>
    public Task<string> SubscribeEventsAsync(Action<string> onChange, CancellationToken token = default) =>
        Api.State.SubscribeStorageAsync(null, (_, change) => onChange(change.ToString() ?? string.Empty), token);

    public async ValueTask DisposeAsync()
    {
        if (Api.IsConnected) await Api.CloseAsync();
        Api.Dispose();
    }
}
